File: CityBikesKit/Station.m
//
//  Station.m
//  BikeCompass
//
//  Created by Raúl Riera on 26/08/2014.
//  Copyright (c) 2014 Raúl Riera. All rights reserved.
//

#import "Station.h"

@implementation Station

#pragma mark - MTLJSONSerializing

+ (NSDictionary *)JSONKeyPathsByPropertyKey
{
    return @{
             @"id": @"id",
             @"name": @"name",
             @"emptySlots": @"empty_slots",
             @"numberOfBikes": @"free_bikes",
             @"latitude": @"latitude",
             @"longitude": @"longitude",
             @"address": @"extra.address"
             };
}

- (BOOL)hasBikes
{
    return self.numberOfBikes > 0;
}

- (NSInteger)totalSlots
{
    return self.numberOfBikes + self.emptySlots;
}

- (BOOL)isEqual:(id)object
{
    if (self == object) {
        return YES;
    } else if (![object isKindOfClass:[Station class]]) {
        return NO;
    } else {
        Station *station = object;
        return [self.id isEqualToString:station.id];
    }
}

- (void)setNilValueForKey:(NSString *)key {
    if ([key isEqualToString:@"emptySlots"]) {
        self.emptySlots = 0;
    } else if ([key isEqualToString:@"numberOfBikes"]) {
        self.numberOfBikes = 0;
    } else {
        [super setNilValueForKey:key];
    }
}

@end

File: App/CityBikesKit/Station.swift
//
//  Station.swift
//  Bike Compass
//
//  Created by Raúl Riera on 23/04/2016.
//  Copyright © 2016 Raul Riera. All rights reserved.
//

import Foundation
import Decodable
import CoreLocation

public struct Station {
    public let id: String
    public let name: String
    public let coordinates: CLLocationCoordinate2D
    public let emptySlots: Int
    public let numberOfBikes: Int
    public let lastUpdated: Date?
}

extension Station: Decodable {
    public static func decode(_ j: AnyObject) throws -> Station {
        return try Station(
            id: j => "id",
            name: j => "name",
            coordinates: CLLocationCoordinate2D(latitude: j => "latitude", longitude: j => "longitude"),
            emptySlots: j => "empty_slots",
            numberOfBikes: j => "free_bikes",
            lastUpdated: try? Date.decode(j => "timestamp")
        )
    }
}

extension Station: Encodable {
    public func encode() -> AnyObject {
        return ["id": id,
                "name": name,
                "latitude": coordinates.latitude,
                "longitude": coordinates.longitude,
                "empty_slots": emptySlots,
                "free_bikes": numberOfBikes]
    }
}
